src/node_crypto.cc

```diff
@@ -2210,52 +2242,91 @@ void Connection::Initialize(Environment* env, Handle<Object> target) {
-  //   VerifyCallback returns 1, the verification process is continued. If
-  //   VerifyCallback always returns 1, the TLS/SSL handshake will not be
-  //   terminated with respect to verification failures and the connection
-  //   will be established. The calling process can however retrieve the
-  //   error code of the last verification error using
-  //   SSL_get_verify_result(3) or by maintaining its own error storage
-  //   managed by VerifyCallback.
-  //
-  //   If no VerifyCallback is specified, the default callback will be
-  //   used.  Its return value is identical to preverify_ok, so that any
-  //   verification failure will lead to a termination of the TLS/SSL
-  //   handshake with an alert message, if SSL_VERIFY_PEER is set.
-  //
-  // Since we cannot perform I/O quickly enough in this callback, we ignore
-  // all preverify_ok errors and let the handshake continue. It is
-  // imparative that the user use Connection::VerifyError after the
-  // 'secure' callback has been made.
+inline int compar(const void* a, const void* b) {
+  return memcmp(a, b, CNNIC_WHITELIST_HASH_LEN);
+}
+
+
+inline int IsSelfSigned(X509* cert) {
+  return X509_NAME_cmp(X509_get_subject_name(cert),
+                       X509_get_issuer_name(cert)) == 0;
+}
+
+
+inline X509* FindRoot(STACK_OF(X509)* sk) {
+  for (int i = 0; i < sk_X509_num(sk); i++) {
+    X509* cert = sk_X509_value(sk, i);
+    if (IsSelfSigned(cert))
+      return cert;
+  }
+  return nullptr;
+}
+
+
+// Whitelist check for certs issued by CNNIC. See
+// https://blog.mozilla.org/security/2015/04/02
+// /distrusting-new-cnnic-certificates/
+inline int IsWhitelistedServerCertificate(X509_STORE_CTX* ctx) {
+  unsigned char hash[CNNIC_WHITELIST_HASH_LEN];
+  unsigned int hashlen = CNNIC_WHITELIST_HASH_LEN;
+
+  STACK_OF(X509)* chain = X509_STORE_CTX_get1_chain(ctx);
+  CHECK_NE(chain, nullptr);
+  CHECK_GT(sk_X509_num(chain), 0);
+
+  // Take the last cert as root at the first time.
+  X509* root_cert = sk_X509_value(chain, sk_X509_num(chain)-1);
+  X509_NAME* root_name = X509_get_subject_name(root_cert);
+
+  if (!IsSelfSigned(root_cert)) {
+    root_cert = FindRoot(chain);
+    CHECK_NE(root_cert, nullptr);
+    root_name = X509_get_subject_name(root_cert);
+  }
+
+  // When the cert is issued from either CNNNIC ROOT CA or CNNNIC EV
+  // ROOT CA, check a hash of its leaf cert if it is in the whitelist.
+  if (X509_NAME_cmp(root_name, cnnic_name) == 0 ||
+      X509_NAME_cmp(root_name, cnnic_ev_name) == 0) {
+    X509* leaf_cert = sk_X509_value(chain, 0);
+    int ret = X509_digest(leaf_cert, EVP_sha256(), hash,
+                          &hashlen);
+    CHECK(ret);
+
+    void* result = bsearch(hash, WhitelistedCNNICHashes,
+                           ARRAY_SIZE(WhitelistedCNNICHashes),
+                           CNNIC_WHITELIST_HASH_LEN, compar);
+    if (result == nullptr) {
+      X509_STORE_CTX_set_error(ctx, X509_V_ERR_CERT_REVOKED);
```

Comment:
One solution is to make this function return an `enum { ALLOWED, DISALLOWED, REVOKED }` instead of an int.  Then, in `VerifyCallback()`, call `X509_STORE_CTX_set_error()` iff the return value is `REVOKED`.

If you change the return value, you should rename the function, because it's no longer the boolean function that the `IsFoo()` name implies.